Add samples for Widgets

thagikura opened the pull request:
Demonstrates various features for Widgets introduced in Android 12 that includes:

- preview layout
- Deferred Configurability
- Widget description
- Animation (smooth transition from Widget to app)
- Recommended widget sizes
- Flexible layouts
- Dynamic coloring
- Simplified RemoteView collections
- Rounded padding and corners

Note that all commits were already reviewed internally and basically the same.

chrisbanes reviewed (approved):
I only took a very quick pass through but LGTM. I've already reviewed all of the changes internally anyway.

yaraki commented:
Should this be merged into the `dev-androidS` branch instead?

thagikura commented:
@yogishadixit I couldn't add you as a reviewer for some reason, but feel free to leave comments in the PR.

yogishadixit commented on AppWidget/app/src/main/java/com/example/android/appwidget/GroceryListAppWidget.kt:
```diff
@@ -0,0 +1,105 @@
+                }
+        }
+    appWidgetManager.updateAppWidget(appWidgetId, remoteViews)
+}
```
I think these should be defined in a companion object.

yogishadixit commented on AppWidget/app/src/main/java/com/example/android/appwidget/ItemsCollectionAppWidget.kt:
```diff
@@ -0,0 +1,184 @@
+/*
```
Can this widget just be merged into the grocery list widget above? Since the grocery widget also has Checkboxes, it should be sufficient to demonstrate the new compound buttons APIs

thagikura replied:
This Widget demonstrates other features other than Checkboxes such as Radio Button, Switch, and simplified usage of ListView. 
Given that, I think it's more readable to have a separate Widget than combining them all.

yogishadixit commented on AppWidget/README.md:
```diff
@@ -0,0 +1,331 @@
+  In API level 31, the colors of your widget will be dynamically determined by the system colors.
+  You should use the system's default theme in order to apply the dynamic coloring (`Theme.DeviceDefault` and
+  `Theme.DeviceDefault.DayNight` for v31.
+  into account).
```
I'm confused by what this means?

thagikura replied:
Sorry,  a wrong sentence was added. Rephrased it.

yogishadixit commented on AppWidget/README.md:
```diff
@@ -0,0 +1,331 @@
+
+- **Padding and rounded corners**
+
+  Following system attributes are introduced in API level 31 for widget paddings and
```
Could we please delete this since it is no longer relevant?

thagikura replied:
Good point. Removed the part mentioning about padding attribute. (attributes for rounded corners are still valid)

yogishadixit commented on AppWidget/README.md:
```diff
@@ -0,0 +1,331 @@
+
+- **Deferrable configurability**
+
+  In API level 31, users are able to reconfigure widgets after they are addd to the home screen
```
```suggestion
  In API level 31, users are able to reconfigure widgets after they are added to the home screen
```

thagikura replied:
Done.

thagikura commented:
Thanks @yogishadixit . Since I have an outstanding CL internally, I addressed your feedback in that CL.

I appreciate if you have a look at the internal CL. I'll add you as a reviewer.

steviek commented on AppWidget/app/src/main/java/com/example/android/appwidget/GroceryListAppWidget.kt:
```diff
@@ -0,0 +1,105 @@
+
+private const val REQUEST_CODE_OPEN_ACTIVITY = 1
+
+@SuppressLint("RemoteViewLayout")
```
I do hope to fix this lint by the time S rolls around, not sure if it should be included here or not based off that.

thagikura replied:
At least with DP3, the use of Checkbox was warned by lint (although it was valid in API level 31)

yogishadixit commented on AppWidget/app/src/main/res/layout/widget_items_collection.xml:
```diff
@@ -0,0 +1,38 @@
+    android:orientation="vertical"
+    android:theme="@style/Theme.AppWidget.AppWidgetContainer">
+
+    <TextView
```
I'd recommend re-using AppWidget/app/src/main/res/layout/widget_grocery_list_title_region.xml here and simply setting the title text when constructing the RemoteViews

thagikura replied:
Changing the text of the TextView requires passing a variable to the include tag, which needs data binding. And I think it adds extract complexity than the benefit of reusing the layout.

yogishadixit replied:
Re-opening

thagikura replied:
This layout doesn't use the ImageButton.

Also, since This is a single TextView, I think having this TextView doesn't make the layout too complex.

yogishadixit commented on AppWidget/app/src/main/res/values/themes.xml:
```diff
@@ -0,0 +1,46 @@
+<resources xmlns:tools="http://schemas.android.com/tools">
+    <!-- Base application theme. -->
+    <style name="Theme.AppWidget" parent="Theme.MaterialComponents.DayNight.DarkActionBar">
+        <!-- Primary brand color. -->
```
I don't think we should explicitly define these colors? Seems like it goes against the concept of dynamic coloring.

thagikura replied:
Although they are not used for widgets, having these set of colors are recommended from material point of view. And at least they are used in the Activity (the status bar color).

I think having them is still better.

yogishadixit commented on AppWidget/app/src/main/java/com/example/android/appwidget/ItemsCollectionAppWidget.kt:
```diff
@@ -0,0 +1,185 @@
+class ItemsCollectionRemoteViewsFactory(private val context: Context) :
+    RemoteViewsService.RemoteViewsFactory {
+
+    companion object {
```
According to the Kotlin style guide, the companion object should be at the end of the class defintion: https://kotlinlang.org/docs/coding-conventions.html#class-layout

thagikura replied:
Done.

yogishadixit commented on AppWidget/app/src/main/java/com/example/android/appwidget/ItemsCollectionAppWidget.kt:
```diff
@@ -0,0 +1,185 @@
+            return builder.setHasStableIds(true).setViewTypeCount(items.count()).build()
+        }
+
+        internal fun getRemoteViewsForLayoutId(
```
constructRemoteViews

thagikura replied:
Done

yogishadixit commented on AppWidget/app/src/main/java/com/example/android/appwidget/ItemsCollectionAppWidget.kt:
```diff
@@ -0,0 +1,185 @@
+            @LayoutRes layoutId: Int
+        ): RemoteViews {
+            val remoteViews = RemoteViews(context.packageName, layoutId)
+            if (BuildCompat.isAtLeastS()) {
```
I think this could benefit from an early return:

```
if (!BuildCompat.isAtLeastS()) {
  return remoteViews;
}
...
```

thagikura replied:
Done.

yogishadixit commented on AppWidget/app/src/main/java/com/example/android/appwidget/ItemsCollectionAppWidget.kt:
```diff
@@ -0,0 +1,185 @@
+import androidx.annotation.LayoutRes
+import androidx.annotation.RequiresApi
+import androidx.core.os.BuildCompat
+import com.example.android.appwidget.ItemsCollectionRemoteViewsFactory.Companion.EXTRA_VIEW_ID
```
Do these need to be imported? Can't they be accessed directly?

thagikura replied:
These need to be imported as it's from another class.

yogishadixit commented on AppWidget/app/src/main/java/com/example/android/appwidget/ItemsCollectionAppWidget.kt:
```diff
@@ -0,0 +1,185 @@
+                    }
+                    R.layout.item_switches -> {
+                        val viewId = R.id.item_switch
+                        val onCheckedChangePendingIntent = PendingIntent.getBroadcast(
```
I think we should register a on checked change listener for the checkboxes and radio buttons as well. The `RemoteViews.setRadioGroupChecked()` and `RemoteViews.setCompoundButtonChecked()` APIs are fairly straightforward, so I think it would be more beneficial to demonstrate how to listen for state changes.

thagikura replied:
Let me do that in a follow up pull request since landing this pull request is time bound. And we are demonstrating how to listen for the state change.

yogishadixit commented on AppWidget/app/src/main/java/com/example/android/appwidget/ItemsCollectionAppWidget.kt:
```diff
@@ -0,0 +1,185 @@
+        }
+    }
+
+    override fun onCreate() {}
```
Potentially silly question: Do these need to be overridden? We're not actually implementing any functionality in these method overrides.

thagikura replied:
They are functions from RemoteViewsFactory. So overriding these are needed.

thagikura commented on AppWidget/app/src/main/java/com/example/android/appwidget/ItemsCollectionAppWidget.kt:
```diff
@@ -0,0 +1,184 @@
+        val remoteViews = RemoteViews(context.packageName, R.layout.widget_items_collection)
+        if (BuildCompat.isAtLeastS()) {
+            val collectionItems =
+                ItemsCollectionRemoteViewsFactory.getRemoteCollectionItems(context)
```
Done.

thagikura commented on AppWidget/app/src/main/java/com/example/android/appwidget/ItemsCollectionAppWidget.kt:
```diff
@@ -0,0 +1,185 @@
+        if (BuildCompat.isAtLeastS() &&
+            intent?.extras?.getInt(REQUEST_CODE) == REQUEST_CODE_FROM_COLLECTION_WIDGET
+        ) {
+            intent.extras?.let {
```
Done.

thagikura commented on AppWidget/app/src/main/res/layout/activity_widget_configure.xml:
```diff
@@ -0,0 +1,107 @@
+    android:padding="16dp">
+
+    <LinearLayout
+        android:id="@+id/grocery_list_container"
```
I think the current layout is better aligned with the internal deck.

yogishadixit replied:
It's causing too much duplication of code, which I worry isn't a good example to set. I think in this case, it's OK if we deviate a little bit from the deck.

thagikura replied:
I'd say it's not too much duplication since each of the list consists of one LinearLayout and three checkboxes.
And I think the appearance looks better with the current layout.

But I extracted some of the common attributes of Checkboxes to style to reduce the duplication.

thagikura commented on AppWidget/app/src/main/res/layout/item_radio_buttons.xml:
```diff
@@ -0,0 +1,41 @@
+    android:theme="@style/Theme.AppWidget.AppWidgetContainer">
+
+    <RadioButton
+        android:id="@+id/item_radio_button1"
```
Even if ID's are not used, I think having ID's doesn't have negative effect.

thagikura commented on AppWidget/app/src/main/res/layout/item_radio_buttons.xml:
```diff
@@ -0,0 +1,41 @@
+
+    <RadioButton
+        android:id="@+id/item_radio_button1"
+        android:layout_width="wrap_content"
```
Done.

thagikura commented on AppWidget/app/src/main/res/layout/item_radio_buttons.xml:
```diff
@@ -0,0 +1,41 @@
+        android:id="@+id/item_radio_button1"
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:buttonTint="?android:attr/colorAccent"
```
Extracted common attributes except for layout_*.

thagikura commented on AppWidget/app/src/main/res/layout/item_switches.xml:
```diff
@@ -0,0 +1,32 @@
+
+    <Switch
+        android:id="@+id/item_switch"
+        android:layout_width="wrap_content"
```
Done.

thagikura commented on AppWidget/app/src/main/res/layout/widget_grocery_grid.xml:
```diff
@@ -0,0 +1,83 @@
+        android:orientation="horizontal">
+
+        <LinearLayout
+            android:id="@+id/todo_grid_left_row"
```
As I wrote it in another comment, having ID's doesn't have negative impact. Also in this case it benefits that having this ID indicates what this LinearLayout is used for.

thagikura commented on AppWidget/app/src/main/res/layout/widget_grocery_list_title_region.xml:
```diff
@@ -0,0 +1,43 @@
+            android:text="@string/grocery_list"
+            android:textColor="?android:attr/textColorPrimary" />
+
+        <ImageButton
```
It demonstrates how the color changes based on the background of the system UI color.
So having this is still beneficial.

thagikura commented on AppWidget/app/src/main/res/layout/widget_todo_list.xml:
```diff
@@ -0,0 +1,66 @@
+    android:orientation="vertical"
+    android:theme="@style/Theme.AppWidget.AppWidgetContainer">
+
+    <LinearLayout
```
Done.

thagikura commented on AppWidget/app/src/main/res/values-v21/styles.xml:
```diff
@@ -0,0 +1,27 @@
+  ~ See the License for the specific language governing permissions and
+  ~ limitations under the License.
+  -->
+<resources>
```
Done.

thagikura commented on AppWidget/app/src/main/AndroidManifest.xml:
```diff
@@ -0,0 +1,74 @@
+
+            <meta-data
+                android:name="android.appwidget.provider"
+                android:resource="@xml/app_widget_info_grocery_list" />
```
Done

thagikura commented on AppWidget/app/src/main/AndroidManifest.xml:
```diff
@@ -0,0 +1,74 @@
+            android:exported="true">
+            <intent-filter>
+                <action android:name="android.intent.action.MAIN" />
+
```
Done.

thagikura commented on AppWidget/app/src/main/java/com/example/android/appwidget/TodoListSharedPrefsUtil.kt:
```diff
@@ -0,0 +1,46 @@
+        appWidgetId: Int,
+        @LayoutRes layoutId: Int
+    ) {
+        context.getSharedPreferences(PREFS_NAME, 0).edit {
```
I think it's better to use named arguments. Introduced private extension function.

thagikura commented:
Thanks @yogishadixit . I think I addressed all (and leave one to a follow up PR).

Since this pull request is time bound, I'd like to merge this hopefully today unless blocking issues are found.
Please take another look.

yogishadixit reviewed (approved):
Looks good overall! Would you just mind running over the code and checking the formatting? Thanks

thagikura commented on AppWidget/app/src/main/java/com/example/android/appwidget/ListAppWidget.kt:
```diff
@@ -0,0 +1,113 @@
+            fun constructRemoteViews(
+                @LayoutRes widgetLayoutId: Int
+            ) = RemoteViews(context.packageName, widgetLayoutId).apply {
+                if (widgetLayoutId == R.layout.widget_grocery_list || widgetLayoutId == R.layout.widget_grocery_grid) {
```
Done.

thagikura commented on AppWidget/app/src/main/java/com/example/android/appwidget/ListAppWidget.kt:
```diff
@@ -0,0 +1,113 @@
+                // Specify the maximum width and height in dp and a layout, which you want to use
+                // for the specified size
+                val viewMapping = mapOf(
+                    SizeF(150f, 110f) to constructRemoteViews(
```
The line from SizeF to the end of constructRemoteViews can't be in a single line.

In that case, I think putting SizeF at the first of the line is more readable to give the same indent to two of them. (Also reformatting suggests the SizeF at the start of the line)
